analyzers/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs

```diff
@@ -58,10 +67,17 @@ protected sealed override TokenTypeInfo CreateMessage(SyntaxTree syntaxTree, Sem
             return tokenTypeInfo;
         }
 
+        protected override void Initialize(SyntaxTree tree) =>
+            SkipIdentifierTokens = HasTooManyIdentifierTokens(tree, identifierTokenKind);
+
+        private static bool HasTooManyIdentifierTokens(SyntaxTree syntaxTree, int tokenKind) =>
+            syntaxTree.GetRoot().DescendantTokens().Count(token => token.RawKind == tokenKind) > IdentifierTokenCountThreshold;
```

Comment:
```suggestion
```
We can avoid all of this and this additional `DescendantTokens()` with better state propagation - see `CreateMessage` above.
It already gathers all tokens. We can do this
```
var skipIdentifierTokens = tokens.Count(token => token.RawKind == tokenKind) > IdentifierTokenCountThreshold
foreach(...
    spans.AddRange(GetTokenClassifier(token, semanticModel, skipIdentifierTokens).Spans);
```
Like that, we can remove `Initialize`, `HasTooManyIdentifierTokens` and `SkipIdentifierTokens`

Reply:
FYI: We need to iterate `DescendantTokens()` two times: once to know how many tokens there are and a second pass to create the spans.

Reply:
Yes, and `DescendantTokens` builds the list from walking the entire tree, while the identifier count would be iteration over flat array.